src/Interpreters/AsynchronousInsertQueue.cpp

```diff
@@ -1097,28 +1106,52 @@ try
 
     try
     {
-        Chunk chunk;
+        size_t num_chunks
+            = key.data_kind == AsynchronousInsertQueueDataKind::Parsed && parse_pool_ptr ? parse_pool_ptr->getMaxThreads() : 1;
+        Chunks chunks(num_chunks);
+
         auto header = pipeline.getSharedHeader();
 
         if (key.data_kind == AsynchronousInsertQueueDataKind::Parsed)
-            chunk = processEntriesWithParsing(key, data, *header, insert_context, log, add_entry_to_asynchronous_insert_log);
+        {
+            if (parse_pool_ptr)
+                processEntriesWithAsyncParsing(chunks, key, data, *header, insert_context, log, add_entry_to_asynchronous_insert_log);
+            else
+                processEntriesWithParsing(chunks, key, data, *header, insert_context, log, add_entry_to_asynchronous_insert_log);
+        }
         else
-            chunk = processPreprocessedEntries(data, *header, insert_context, add_entry_to_asynchronous_insert_log);
+        {
+            processPreprocessedEntries(chunks, data, *header, insert_context, add_entry_to_asynchronous_insert_log);
+        }
 
-        ProfileEvents::increment(ProfileEvents::AsyncInsertRows, chunk.getNumRows());
+        size_t num_rows = 0;
+        size_t num_bytes = 0;
 
-        if (chunk.getNumRows() == 0)
+        for (auto & achunk : chunks)
+        {
+            num_rows += achunk.getNumRows();
+            num_bytes += achunk.bytes();
+        }
+
+        if (num_rows == 0)
         {
             pipeline.cancel(); // this just cancels the processors
             finish_entries(std::move(pipeline), /*num_rows=*/ 0, /*num_bytes=*/ 0);
             return;
         }
 
-        size_t num_rows = chunk.getNumRows();
-        size_t num_bytes = chunk.bytes();
+        if (parse_pool_ptr)
+        {
+            auto source = std::make_unique<SourceFromChunks>(header, std::move(chunks));
```

Comment:
This has big impact.
Before there was on chunk, now there are many chunks. 
Before it has been inserted as one part in landing table,
not it could be as many parts as chunks count here.

Reply:
That's a very good concern, thanks. 
I'll recheck if it is true, and if it is, rollback to a previous version where I joined chunks explicitly.

Reply:
>> What is wrong with AsynchronousInsertQueue::processData?

> Do you suggest joining parts in AsynchronousInsertQueue::processData ?
Could you provide more reasons/hints?

It was the question on your statement

> Seems slightly better than doing basically that same in AsynchronousInsertQueue::processData .